File: solutions/1448.count-good-nodes-in-binary-tree.java
/*
 * @lc app=leetcode id=1448 lang=java
 *
 * [1448] Count Good Nodes in Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int goodNodesCounter = 0;

    public int goodNodes(TreeNode root) {
        checkNodes(root.val, root);
        return goodNodesCounter;
    }

    private void checkNodes(int maximum, TreeNode node) {
        if (node == null) return;

        if (node.val >= maximum) {
            goodNodesCounter++;
        }

        int nextMax = Math.max(maximum, node.val);
        checkNodes(nextMax, node.left);
        checkNodes(nextMax, node.right);
    }
}
// @lc code=end

File: solutions/1448.count-good-nodes-in-binary-tree.py
#
# @lc app=leetcode id=1448 lang=python3
#
# [1448] Count Good Nodes in Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        good_nodes = 0

        def checkNodes(node, maximum):
            nonlocal good_nodes
            if not node:
                return

            if node.val >= maximum:
                good_nodes += 1

            maximum = max(maximum, node.val)
            checkNodes(node.left, maximum)
            checkNodes(node.right, maximum)

        checkNodes(root, root.val)
        return good_nodes

# @lc code=end

